source/common/json/json_loader.h

```diff
@@ -20,6 +27,186 @@ class Factory {
+    ObjectPtr getObject(const std::string& name, bool allow_empty) const override;
+    std::vector<ObjectPtr> getObjectArray(const std::string& name) const override;
+    std::string getString(const std::string& name) const override;
+    std::string getString(const std::string& name, const std::string& default_value) const override;
+    std::vector<std::string> getStringArray(const std::string& name) const override;
+    std::vector<ObjectPtr> asObjectArray() const override;
+    std::string asString() const override { return stringValue(); }
+
+    bool empty() const override;
+    bool hasObject(const std::string& name) const override;
+    void iterate(const ObjectCallback& callback) const override;
+    void validateSchema(const std::string& schema) const override;
+
+  private:
+    enum class Type {
+      Array,
+      Boolean,
+      Double,
+      Integer,
+      Null,
+      Object,
+      String,
+    };
+
+    struct Value {
+      std::vector<FieldPtr> array_value_;
+      bool boolean_value_;
+      double double_value_;
+      int64_t integer_value_;
+      std::map<const std::string, FieldPtr> object_value_;
+      std::string string_value_;
+    };
+
+    explicit Field(Type type) : type_(type) {}
+    explicit Field(const std::string& value) : type_(Type::String) { value_.string_value_ = value; }
+    explicit Field(int64_t value) : type_(Type::Integer) { value_.integer_value_ = value; }
+    explicit Field(double value) : type_(Type::Double) { value_.double_value_ = value; }
+    explicit Field(bool value) : type_(Type::Boolean) { value_.boolean_value_ = value; }
+
+    bool isType(Type type) const { return type == type_; }
+    void checkType(Type type) const {
+      if (!isType(type)) {
+        throw Exception("Field access in JSON with correct type.");
+      }
+    }
+    // value rtype funcs
+    std::string stringValue() const {
+      checkType(Type::String);
+      return value_.string_value_;
+    }
+    std::vector<FieldPtr> arrayValue() const {
+      checkType(Type::Array);
+      return value_.array_value_;
+    }
+    bool booleanValue() const {
+      checkType(Type::Boolean);
+      return value_.boolean_value_;
+    }
+    double doubleValue() const {
+      checkType(Type::Double);
+      return value_.double_value_;
+    }
+    int64_t integerValue() const {
+      checkType(Type::Integer);
+      return value_.integer_value_;
+    }
+
+    rapidjson::Document asRapidJsonDocument() const;
+    static void build(const Field& field, rapidjson::Value& value,
+                      rapidjson::Document::AllocatorType& allocator);
+
+    uint64_t line_number_start_;
+    uint64_t line_number_end_;
+    Type type_;
+    Value value_;
+  };
+
+  class LineCountingStringStream : public rapidjson::StringStream {
+    // Ch is typdef in parent class to handle character encoding.
+  public:
+    LineCountingStringStream(const Ch* src) : rapidjson::StringStream(src), line_number_(1) {}
+    Ch Take() {
```

Comment:
is that an override of method from rapidjson::StringStream? (if so Ch Take() override)